### PR TITLE
[#16] Fix race condition when fetching OIDC document

### PR DESCRIPTION
**Pull request by ieugen**
This is my attempt at fixing #16 . 
It does work locally but can't deploy it easily since traefik does not support fork. 
https://community.traefik.io/t/does-the-plugin-directory-support-github-forks/18462

I could hard fork but then it will make it harder to contribute changes back to upstream. 
I imagine you are not a big fan of getting git patches over email. 

I would appreciate it very much if you can merge this sooner rather than later.
I have a service that is down for a few days now because of it. 

Please :) .

Thanks,
Eugen

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -20,7 +20,55 @@ type TraefikOidcAuth struct {
 	Jwks              *JwksHandler
 }
 
+// Make sure we fetch oidc discovery document during first request - avoid race condition
+// Perform lock when changing document - we are in concurrent environment
+func EnsureOidcDiscovery(toa *TraefikOidcAuth) (*TraefikOidcAuth, error) {
+	var config = toa.Config
+	var parsedURL = toa.ProviderURL
+	if toa.DiscoveryDocument == nil {
+		var jwks = &JwksHandler{}
```
Comment (sevensolutions) [inline main.go:29]: I think we shouldn't mix the JWKS lock with the discovery document lock.
I'am not a Go-expert, but if we create a new JwksHandler here, then i think we will have a new lock on every call, which makes it useless.
Maybe we create another mutex on the "toa" (TraefikOidcAuth) for getting the discovery document.

Comment (sevensolutions) [reply main.go:29]: And: I think after we've acquired the lock, we should check for `if toa.DiscoveryDocument == nil` again.

File: config.go [hunk 2]
```diff
@@ -156,32 +155,13 @@ func New(uctx context.Context, next http.Handler, config *Config, name string) (
...
-		ProviderURL:       parsedURL,
-		Config:            config,
-		DiscoveryDocument: oidcDiscoveryDocument,
-		Jwks: &JwksHandler{
-			Url: oidcDiscoveryDocument.JWKSURI,
-		},
+		next:        next,
+		ProviderURL: parsedURL,
+		Config:      config,
+		// DiscoveryDocument: oidcDiscoveryDocument,
```
Comment (sevensolutions) [inline config.go:162]: Can you please remove these comments completely?

**Review by sevensolutions — CHANGES_REQUESTED** [review]
@ieugen thx for your contribution.
Can you please have a look on my comments?
I will then fetch your changes and also do some tests.

Yeah using a fork doesn't work but you can temporarily copy the entire repo to your server and reference it. This should work.
But once I've merged your PR i can push a new tag and release it as a fix immediately.

File: config.go (hunk 2 above)
Comment (ieugen) [reply config.go:162]: yes, done.

File: main.go (hunk 1 above)
Comment (ieugen) [reply main.go]: Fixed both of them.
I am creating a JwksHandler since it's needed .

**Comment by ieugen** [comment]
I tested the setup locally with kanidm - I had the setup ready.

**Comment by sevensolutions** [comment]
@ieugen Thx, works great. I just fetched your PR and also tested it against ZITADEL.
Can i do just two more small cosmetic changes to your PR?
I just want to try that because i've never done that. I found this guide on how to do it:
https://tighten.com/insights/adding-commits-to-a-pull-request/

**Pull request merged**